### PR TITLE
Six extra seconds of cooperation

### PR DESCRIPTION
**Pull request by maximumpower55**

### CODE REVIEW

File: src/main/java/io/github/fusionflux/portalcubed/mixin/client/FogRendererMixin.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(FogRenderer.class)
+public class FogRendererMixin {
+	@Shadow private static float fogRed;
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/mixin/client/FogRendererMixin.java:23]: put annotations on separate lines

File: src/main/java/io/github/fusionflux/portalcubed/mixin/client/EntityRenderDispatcherMixin.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	@ModifyArgs(
+			method = "render",
+			at = @At(
+					value = "INVOKE",
+					target = "Lnet/minecraft/client/renderer/entity/EntityRenderer;render(Lnet/minecraft/world/entity/Entity;FFLcom/mojang/blaze3d/vertex/PoseStack;Lnet/minecraft/client/renderer/MultiBufferSource;I)V"
+			)
+	)
+	private void disintegrationRendering(
+			Args args,
+			@Local(argsOnly = true) Entity entity,
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/mixin/client/EntityRenderDispatcherMixin.java:31]: don't use local for this, just add the method args to the method

File: src/main/java/io/github/fusionflux/portalcubed/mixin/client/EntityRenderDispatcherMixin.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+			@Local(ordinal = 1, argsOnly = true) float tickDelta,
+			@Local(argsOnly = true) MultiBufferSource bufferSource
+	) {
+		if (entity.pc$disintegrating()) {
+			args.set(2, 0f);
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/mixin/client/EntityRenderDispatcherMixin.java:36]: comment what this is

File: src/main/java/io/github/fusionflux/portalcubed/mixin/client/ClientLevelMixin.java [hunk 4]
```diff
@@ -45,6 +49,15 @@ private void init(CallbackInfo ci) {
 		this.portalManager = new ClientPortalManager((ClientLevel) (Object) this);
 	}
 
+	@WrapOperation(method = "tickNonPassenger", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/entity/Entity;tick()V"))
+	private void disintegrationTick(Entity instance, Operation<Void> original) {
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/mixin/client/ClientLevelMixin.java:53]: ticking is weird, make sure passengers disintegrate properly

File: src/main/java/io/github/fusionflux/portalcubed/mixin/EntityMixin.java [hunk 5]
```diff
@@ -23,26 +42,182 @@
...
+	@Unique
+	private int disintegrateTicks;
+
+	@Override
+	public boolean pc$disintegrate() {
+		if (!this.disintegrating) {
+			this.disintegrating = true;
+			if (this.level() instanceof ServerLevel && (Object) this instanceof Entity self) {
+				BlockState feetState = this.getFeetBlockState();
+				if (feetState.getBlock() instanceof FloorButtonBlock floorButton && floorButton.isEntityPressing(feetState, this.blockPosition(), self))
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/mixin/EntityMixin.java:116]: comment why this is done

File: src/main/java/io/github/fusionflux/portalcubed/framework/extension/EntityExt.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package io.github.fusionflux.portalcubed.framework.extension;
+
+public interface EntityExt {
+	int DISINTEGRATE_TICKS = 3 * 20;
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/framework/extension/EntityExt.java:4]: I don't like storing these here

File: src/main/java/io/github/fusionflux/portalcubed/content/prop/PropRenderer.java [hunk 7]
```diff
@@ -43,6 +46,7 @@ public void render(Prop prop, float yaw, float tickDelta, PoseStack matrices, Mu
 		matrices.popPose();
 	}
 
+	@UnknownNullability
 	@Override
 	public ResourceLocation getTextureLocation(Prop entity) {
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/content/prop/PropRenderer.java:51]: this shouldn't return null, use missingno

File: src/main/java/io/github/fusionflux/portalcubed/content/fizzler/DisintegrationSoundType.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+	RADIO,
+	TURRET;
+
+	public final String name = this.name().toLowerCase(Locale.ROOT);
+	public final SoundEvent sound;
+	public final TagKey<EntityType<?>> tag;
+
+	DisintegrationSoundType() {
+		ResourceLocation id = PortalCubed.id(name + "_disintegration");
+		this.sound = Registry.register(BuiltInRegistries.SOUND_EVENT, id, SoundEvent.createVariableRangeEvent(id));
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/content/fizzler/DisintegrationSoundType.java:29]: sound events should be in a PortalCubedSounds and then referenced in constructors

Comment (TropheusJ) [reply src/main/java/io/github/fusionflux/portalcubed/content/fizzler/DisintegrationSoundType.java:29]: similar thing for tags

File: src/main/java/io/github/fusionflux/portalcubed/content/PortalCubedBlocks.java [hunk 9]
```diff
@@ -849,12 +842,22 @@ public class PortalCubedBlocks {
...
+			.settings(s -> s.mapColor(MapColor.TERRACOTTA_GREEN))
+			.item(BlockItemProvider::noItem)
+			.build();
+
+	public static final Block GOO_CAULDRON = REGISTRAR.blocks.createFrom("toxic_goo_cauldron", GooCauldronBlock::new, Blocks.CAULDRON)
+			.item(BlockItemProvider::noItem)
+			.build();
+
 	public static void init() {
+		Item.BY_BLOCK.put(GOO_CAULDRON, Items.CAULDRON);
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/content/PortalCubedBlocks.java:860]: ???
if this is for pickblock, just override the method in the goo cauldron class

File: src/main/java/io/github/fusionflux/portalcubed/PortalCubed.java [hunk 10]
```diff
@@ -51,6 +56,11 @@ public void onInitialize(ModContainer mod) {
 
 		ConstructManager.init();
 		HoldableEntity.registerEventListeners();
+		DisintegrationSoundType.init();
+		EntityTrackingEvents.AFTER_START_TRACKING.register((tracked, player) -> {
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/PortalCubed.java:60]: I don't like putting this here

File: gradle.properties [hunk 11]
```diff
@@ -14,9 +14,12 @@ loader_version = 0.24.0
...
 terraform_wood_api_version = 9.0.0
+sodium_version = 0.5.8
 
 # Dev env mods
 # https://modrinth.com/mod/modmenu/versions
 mod_menu_version = 9.0.0
 # https://modrinth.com/mod/emi/versions
 emi_version = 1.1.4+1.20.4
+# https://modrinth.com/mod/indium/versions
+indium_version = 1.0.30+mc1.20.4
```
Comment (TropheusJ) [inline gradle.properties:25]: this is a dependency, not dev QOL. put it with sodium

Comment (TropheusJ) [reply gradle.properties:25]: it should also have a QMJ dependency

**Review by TropheusJ — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/fusionflux/portalcubed/framework/extension/EntityExt.java (hunk 6 above)
Comment (maximumpower55) [reply src/main/java/io/github/fusionflux/portalcubed/framework/extension/EntityExt.java:4]: Would it be better if EntityExt was renamed to DisintegrationEntityExtensions or something like that

File: src/main/java/io/github/fusionflux/portalcubed/content/PortalCubedBlocks.java (hunk 9 above)
Comment (maximumpower55) [reply src/main/java/io/github/fusionflux/portalcubed/content/PortalCubedBlocks.java:860]: It is for pickblock, mojang did it this way too see `Items.registerBlock(Block block, Block... blocks)` and `Items.CAULDRON`. I guess I could just override the method though

File: src/main/java/io/github/fusionflux/portalcubed/PortalCubed.java (hunk 10 above)
Comment (maximumpower55) [reply src/main/java/io/github/fusionflux/portalcubed/PortalCubed.java:60]: I could put in a static method in `EntityExt` but I would like it to be renamed if so, see my reply to `I don't like storing these here` in `EntityExt`

File: src/main/java/io/github/fusionflux/portalcubed/mixin/client/ClientLevelMixin.java (hunk 4 above)
Comment (maximumpower55) [reply src/main/java/io/github/fusionflux/portalcubed/mixin/client/ClientLevelMixin.java:53]: They don't, but disintegration makes them not passengers (`stopRiding` gets called)

File: src/main/java/io/github/fusionflux/portalcubed/content/fizzler/DisintegrationSoundType.java (hunk 8 above)
Comment (maximumpower55) [reply src/main/java/io/github/fusionflux/portalcubed/content/fizzler/DisintegrationSoundType.java:29]: I don't really like this plus you didn't have any problems with the prop impact sound registering the sound event in its enum constructor

File: src/main/java/io/github/fusionflux/portalcubed/mixin/client/ClientLevelMixin.java (hunk 4 above)
Comment (TropheusJ) [reply src/main/java/io/github/fusionflux/portalcubed/mixin/client/ClientLevelMixin.java:53]: what about something starting riding after fizzling

File: src/main/java/io/github/fusionflux/portalcubed/content/fizzler/DisintegrationSoundType.java (hunk 8 above)
Comment (TropheusJ) [reply src/main/java/io/github/fusionflux/portalcubed/content/fizzler/DisintegrationSoundType.java:29]: I didn't see that

File: src/main/java/io/github/fusionflux/portalcubed/framework/extension/EntityExt.java (hunk 6 above)
Comment (TropheusJ) [reply src/main/java/io/github/fusionflux/portalcubed/framework/extension/EntityExt.java:4]: no. I guess it's fine

File: src/main/java/io/github/fusionflux/portalcubed/PortalCubed.java (hunk 10 above)
Comment (TropheusJ) [reply src/main/java/io/github/fusionflux/portalcubed/PortalCubed.java:60]: there's not really anywhere better so I guess it's fine

File: src/main/java/io/github/fusionflux/portalcubed/content/fizzler/DisintegrationSoundType.java (hunk 8 above)
Comment (TropheusJ) [reply src/main/java/io/github/fusionflux/portalcubed/content/fizzler/DisintegrationSoundType.java:29]: I guess it's fine

File: src/main/java/io/github/fusionflux/portalcubed/content/goo/GooCauldronBlock.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+	@NotNull
+	@Override
+	public MapCodec<GooCauldronBlock> codec() {
+		return CODEC;
+	}
+
+	@NotNull
+	@Override
+	public Item asItem() {
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/content/goo/GooCauldronBlock.java:41]: getCloneItemStack, not asItem

**Pull request merged**